node/src/components/consensus/consensus_protocol.rs

```diff
@@ -152,4 +152,8 @@ pub(crate) trait ConsensusProtocol<I, C: Context> {
 
     /// Returns true if the protocol has received some messages since initialization.
     fn has_received_messages(&self) -> bool;
+
+    /// Returns an iterator over all the values that are expected to become finalized, but are not
+    /// finalized yet.
+    fn non_finalized_values<'a>(&'a self) -> Box<dyn Iterator<Item = &'a C::ConsensusValue> + 'a>;
```

Comment:
Wouldn't it make more sense to include that information in `ProtocolOutcome::CreateNewBlock`?
It also feels safer, because if you call `non_finalized_values` at a point in time where the protcol state has changed since the new block request was generated, it could return values corresponding to a different fork choice.

Reply:
…or at least put the fork choice into `CreateNewBlock`, so the rest of the computation doesn't change anymore if something gets added to the protocol state.

Reply:
> if you call non_finalized_values at a point in time where the protcol state has changed since the new block request was generated, it could return values corresponding to a different fork choice

Isn't that an argument _against_ this change? I mean, you would want to use a fork choice that is as close to the one you will use when proposing the block as possible, right? So if it _does_ change in the meantime, it makes more sense to use the later one rather than the earlier one, no?

Otherwise, including this in `ProtocolOutcome` would avoid modifying the trait, so there's that.

Reply:
When we request deploys, we [already determine the panorama](https://github.com/CasperLabs/casper-node/blob/6fbf835787230d499e7a13066c84cc8b22540851/node/src/components/consensus/highway_core/active_validator.rs#L207) and store it in `ActiveValidator`. We should use the fork choice that corresponds exactly to that panorama.

Reply:
Oh! Okay, then I guess your solution is much better :+1:

Reply:
I'm also wondering if it's worth it to fiddle with returning a boxed iterator, or whether I should just be returning a `Vec` or something instead.